### PR TITLE
[TC-545] Adds the ability to disable traffic routing on a per-profile basis.

### PR DESCRIPTION
**Pull request by dg4prez**
This adds the ability to disable routing to edge caches on a per-profile basis.  Servers whose profiles have this option enabled will no longer show delivery services in the crconfig.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/27/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/157/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/118/

File: docs/source/development/traffic_ops_api/v12/profile.rst [hunk 1]
```diff
@@ -71,6 +73,7 @@ Profiles
             "cdn": 1
             "cdnName": "cdn1"
             "type": "ATS_PROFILE"
+            "routingDisabled": 0
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/profile.rst]: booleans should be returned or submitted as true|false not 1|0. this changed when we migrated to postgres.

Comment (dg4prez) [reply docs/source/development/traffic_ops_api/v12/profile.rst]: That's what they're coming out of the db function as - they are in the DB as booleans.  Should we manually change them to true|false at the json level?

File: traffic_ops/app/db/migrations/20170825124926_ignore_routing.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
+
+-- +goose Up
+-- SQL in section 'Up' is executed when this migration is applied
+
+ALTER TABLE profile
+    ADD ccr_routing_disabled BOOL,
```
Comment (mitchell852) [inline traffic_ops/app/db/migrations/20170825124926_ignore_routing.sql]: how about routing_disabled instead?

Comment (mitchell852) [reply traffic_ops/app/db/migrations/20170825124926_ignore_routing.sql]: maybe you should make it like this:

create a not null column called routing_disabled with default=false

then in the api if routingDisabled is not defined, set it's value to false for the insert/update...

File: traffic_ops/app/db/migrations/20170825124926_ignore_routing.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    ALTER COLUMN ccr_routing_disabled SET DEFAULT FALSE;
+UPDATE profile SET ccr_routing_disabled = FALSE where ccr_routing_disabled = NULL;
```
Comment (mitchell852) [inline traffic_ops/app/db/migrations/20170825124926_ignore_routing.sql]: not sure you will need this UPDATE statement if you do what i said above...

File: traffic_ops/app/lib/API/Profile.pm [hunk 4]
```diff
@@ -43,6 +43,7 @@ sub index {
 					"cdn" 			=> defined($row->profile->cdn) ? $row->profile->cdn->id : undef,
 					"cdnName" 		=> defined($row->profile->cdn) ? $row->profile->cdn->name : undef,
 					"type" 			=> $row->profile->type,
+					"routingDisabled"	=> defined($row->profile->ccr_routing_disabled) ? $row->profile->ccr_routing_disabled : undef,
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: this will always be defined if you make the column not null

File: traffic_ops/app/lib/API/Profile.pm [hunk 5]
```diff
@@ -208,6 +217,7 @@ sub create {
 			description => $description,
 			cdn         => $cdn,
 			type        => $type,
+			ccr_routing_disabled => $ccr_routing_disabled,
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: you can just set this to 

ccr_routing_disabled => defined($params->{routingDisabled}) ? $params->{routingDisabled} : 0;

File: traffic_ops/app/lib/API/Profile.pm [hunk 6]
```diff
@@ -333,6 +344,15 @@ sub update {
 		}
 	}
 
+	if ( !defined($params->{routingDisabled})) {
+		return $self->alert("routingDisabled is required - default is 0");
+	}
+	if ($params->{routingDisabled} != 0 && $params->{routingDisabled} != 1) {
+		return $self->alert("routingDisabled must be 0 or 1. Default is 0.");
+	}
+	print STDERR Dumper("I think I've got an update for routingDisabled");
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: kill this

File: traffic_ops/app/lib/Schema/Result/Cdn.pm [hunk 7]
```diff
@@ -92,7 +92,7 @@ __PACKAGE__->set_primary_key("id");
```
Comment (mitchell852) [inline traffic_ops/app/lib/Schema/Result/Cdn.pm]: you should get rid of all these unrelated schema file changes

File: docs/source/development/traffic_ops_api/v12/profile.rst (hunk 1 above)
Comment (mitchell852) [reply docs/source/development/traffic_ops_api/v12/profile.rst]: look at deliveryservice.pm...lots of booleans coming out of that api

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/60/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/61/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/62/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-whatchanged/63/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/152/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/153/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/154/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/155/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/191/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/192/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/193/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/194/

**Comment by mitchell852** [comment]
@dg4prez - i'm seeing a bunch of broken tests

t/api/1.2/cdn.t                               (Wstat: 512 Tests: 56 Failed: 2)
  Failed tests:  13, 18
  Non-zero exit status: 2
t/api/1.2/config_files_apache_traffic_server.t (Wstat: 256 Tests: 54 Failed: 1)
  Failed test:  50
  Non-zero exit status: 1
t/api/1.2/deliveryservice.t                   (Wstat: 512 Tests: 174 Failed: 2)
  Failed tests:  80, 167
  Non-zero exit status: 2
t/api/1.2/profile.t                           (Wstat: 1792 Tests: 96 Failed: 7)
  Failed tests:  61-67
  Non-zero exit status: 7
t/profile.t                                   (Wstat: 256 Tests: 12 Failed: 1)
  Failed test:  9
  Non-zero exit status: 1

File: traffic_ops/app/lib/API/Profile.pm [hunk 8]
```diff
@@ -190,6 +194,11 @@ sub create {
 		return $self->alert("Profile type is required.");
 	}
 
+	if ( defined( $params->{routingDisabled} ) && $params->{routingDisabled} != (0 || 1) ) {
+		print STDERR Dumper($params->{routingDisabled});
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: you should take out the print statement

File: traffic_ops/app/lib/API/Profile.pm [hunk 9]
```diff
@@ -333,6 +345,13 @@ sub update {
 		}
 	}
 
+	if ( defined( $params->{routingDisabled} ) && $params->{routingDisabled} != (0 || 1) ) {
+		print STDERR Dumper($params->{routingDisabled});
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: this one too

File: traffic_ops/app/lib/API/Profile.pm [hunk 10]
```diff
@@ -333,6 +345,13 @@ sub update {
...
+		return $self->alert("Routing disabled is a boolean value.");
+	}
+
+	my $routing_disabled = defined($params->{routingDisabled}) ? $params->{routingDisabled} : 0;
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: this won't work. if you send thru routingDisabled: true or routingDisabled: false it will always put a 1 in the database. you might need to consult @dangogh (our resident perl expert) on what the best approach is here. 

Also, you might want to retest thru the API, i keep getting this error when i send thru routingDisabled: false

Routing disabled is a boolean value.

Comment (mitchell852) [reply traffic_ops/app/lib/API/Profile.pm]: Also, you need to add some unit tests to test this new column. at a minimum, create and update a profile thru the api with routingDisabled=true and routingDisabled=false

Comment (mitchell852) [reply traffic_ops/app/lib/API/Profile.pm]: Also, can you add your new field to this api endpoint?

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/API/Profile.pm#L87

Comment (mitchell852) [reply traffic_ops/app/lib/API/Profile.pm]: Also, I created a PR for you that adds this field to the Traffic Portal

https://github.com/dg4prez/incubator-trafficcontrol/pull/4

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/285/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/243/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/299/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/258/

File: traffic_ops/app/lib/API/Profile.pm (hunk 10 above)
Comment (dg4prez) [reply traffic_ops/app/lib/API/Profile.pm]: All fixed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/261/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/302/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/304/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/263/

File: traffic_ops/app/lib/API/Profile.pm [hunk 11]
```diff
@@ -222,15 +227,25 @@ sub create {
 	if ($existing_desc) {
 		return $self->alert("a profile with the exact same description already exists.");
 	}
-
+	
 	my $cdn = $params->{cdn};
 	my $type = $params->{type};
+	my $routing_disabled = defined($params->{routingDisabled}) ? $params->{routingDisabled} : 0;
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: @dangogh - is there a better way to do this? I can't imagine there isn't...

Comment (dangogh) [reply traffic_ops/app/lib/API/Profile.pm]: not really when it's being used as dbix input..

btw,  @dg4prez, @mitchell852  unit tests fail on this -- check out https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/263/

File: traffic_ops/app/lib/API/Profile.pm [hunk 12]
```diff
@@ -222,15 +227,25 @@ sub create {
...
+	# Boolean values don't always show properly, so we're going to evaluate these then convert them to standard integers.
+	# This allows the response output to always show true/false correctly.
+	if ($routing_disabled == 1) {
```
Comment (mitchell852) [inline traffic_ops/app/lib/API/Profile.pm]: no @dangogh look at this part. this part makes me want to rethink programming :)

Comment (dg4prez) [reply traffic_ops/app/lib/API/Profile.pm]: I hate it as well, but there's a reason for it. I want the user to be able to use either 1,0,true, or false for the json input, and have it always output true/false in the response.  if the user sends 0 or 1, \$value outputs as true/false but when the user inputs true/false it outputs as a reference.  This basically sets the value to 0 or 1 regardless of the method the user uses in the json.

File: traffic_ops/app/lib/API/Profile.pm (hunk 11 above)
Comment (dg4prez) [reply traffic_ops/app/lib/API/Profile.pm]: I don't understand why it's failing.  Here's the test locally:

kabletown app $ git status
On branch routing_ignore
Your branch is up-to-date with 'origin/routing_ignore'.

nothing to commit, working tree clean
kabletown app $ db/admin.pl --env=test reset
Creating user: traffic_ops
Dropping database: to_test
DROP DATABASE IF EXISTS to_test;
CREATE DATABASE to_test OWNER traffic_ops;
Creating database tables.
blah blah blah db stuff

kabletown app $ prove -qrp t/api/1.2/profile.t
t/api/1.2/profile.t .. Subroutine TrafficOps::has redefined at /opt/traffic_ops/app/local/lib/perl5/Mojo/Base.pm line 38.
Subroutine TrafficOps::has redefined at /opt/traffic_ops/app/local/lib/perl5/Mojo/Base.pm line 38.
NOTICE:  truncate cascades to table "asn"
NOTICE:  truncate cascades to table "cachegroup_parameter"
NOTICE:  truncate cascades to table "server"
NOTICE:  truncate cascades to table "staticdnsentry"
NOTICE:  truncate cascades to table "deliveryservice_server"
NOTICE:  truncate cascades to table "hwinfo"
NOTICE:  truncate cascades to table "servercheck"
t/api/1.2/profile.t .. 16/? DBIx::Class::ResultSet::search_rs(): search( %condition ) is deprecated, use search( \%condition ) instead at /opt/traffic_ops/app/lib/API/Parameter.pm line 137
t/api/1.2/profile.t .. ok
All tests successful.
Files=1, Tests=99,  4 wallclock secs ( 0.03 usr  0.01 sys +  2.41 cusr  0.59 csys =  3.04 CPU)
Result: PASS

Comment (mitchell852) [reply traffic_ops/app/lib/API/Profile.pm]: you need to run All the tests with prove -qrp t

Test Summary Report
-------------------
t/api/1.2/cdn.t                               (Wstat: 512 Tests: 56 Failed: 2)
  Failed tests:  13, 18
  Non-zero exit status: 2
t/api/1.2/config_files_apache_traffic_server.t (Wstat: 256 Tests: 54 Failed: 1)
  Failed test:  50
  Non-zero exit status: 1
t/api/1.2/deliveryservice.t                   (Wstat: 512 Tests: 180 Failed: 2)
  Failed tests:  83, 173
  Non-zero exit status: 2
t/profile.t                                   (Wstat: 256 Tests: 12 Failed: 1)
  Failed test:  9
  Non-zero exit status: 1

Comment (dg4prez) [reply traffic_ops/app/lib/API/Profile.pm]: OK, tests should be fixed.  Waiting on results.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/311/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/270/
